### PR TITLE
Rotate tool upgrades

### LINKED ISSUE #26936: snapping on hover when we want to rotate features depending a specific anchor point
Author Name: **Léo Chevrier** (Léo Chevrier)
Original Redmine Issue: [19106](https://issues.qgis.org/issues/19106)

Redmine category:digitising
Assignee: Léo Chevrier

---
Hello,
One of usefull evolution for the rotate feature's function in qgis would be to snapp on hover on the anchor point that we want. 

On my exemple : I want to rotate a segment with the anchor point on the end of the segment.

So I have to Use _Rotate Feature(s) to rotate one or multiple features in the map canvas. Press the rotateFeature Rotate Feature(s) icon and then click on the feature to rotate. Either click on the map to place the rotated feature or enter an angle in the user input widget. 
If you enable the map tool with feature(s) selected, its (their) centroid appears and will be the rotation anchor point._ *If you want to move the anchor point, hold the Ctrl button and click on the map to place it.* (in this manipulation, i can't snapp my mouse pointers on the end of my segment, I have to manually zoom in on it and approximatively clickin on it)

What do you think about this evolution ? 

Best regards

Léo

---

Related issue(s): #27073 (duplicates)
Redmine related issue(s): [19245](https://issues.qgis.org/issues/19245)

---

### LINKED ISSUE #33353: Snap for rotate tool
**Feature description.**

For accurate rotation element, is necessary:
1.- Set the origin rotation 
2.- Set the rotation angle
Qgis allows to specify accuratly the angle of rotation, but not allow to snap on an existing point to set the origin point rotation.

It would be convenient to implement this possibility.
**Additional context**

### LINKED ISSUE #39698: Rotate tool: Snapping to rotate around vertex and geometry snapping for placement.
The freehand rotate tool could use these extra features to improve the workflow when using it.

1. Snapping to rotate around a vertex is already discussed here  #26936, but I added it for context to the second feature.

2. After choosing the rotate point, another vertex of the geometry being rotated could also be selected on snap, and placed on snap to features in other layers.
The effect is similar to what the Move tool does right now: You can grab a feature by a vertex, and place this feature on snap with any other feature available, leaving it perfectly aligned/adjacent.

It may be specific to what I do, but in my line of work, I have to design certain walking paths inside polygons. These polygons repeat (and therefore the path inside them is the same), but are in different orientations. I have to reorient the path (line feature) by hand every time, since redrawing it is a bit time consuming. Doing it by snapping removes the need to zoom in endlessly to reduce misalignment to the minimum and should be far faster.

Thanks a lot for maintaining this awesome software.

### PR DESCRIPTION
**Pull request by uclaros**
## Description
Upgrades for the *Rotate tool* to support snapping and advanced digitizing toolbar.
Fixes #26936
Fixes #33353
Fixes #39698

### CODE REVIEW

**Comment by roya0045** [comment]
Wondering if scaler would also benefit from snaping, Cad might not be as critical.

**Comment by uclaros** [comment]
Definitely! Just let me know if you are up to it yourself so we don't do duplicate work.

**Comment by roya0045** [comment]
> Definitely! Just let me know if you are up to it yourself so we don't do duplicate work.

Got enough on my hands at the moment. If you can do it, it'll be appreciated. Otherwise we can wait until someone requests that.

File: src/app/qgsmaptoolrotatefeature.cpp [hunk 1]
```diff
@@ -139,24 +141,30 @@ void QgsAngleMagnetWidget::angleSpinBoxValueChanged( double angle )
...
+
   if ( mRotationActive )
   {
-    const double XDistance = e->pos().x() - mStPoint.x();
-    const double YDistance = e->pos().y() - mStPoint.y();
-    double rotation = std::atan2( YDistance, XDistance ) * ( 180 / M_PI ) - mRotationOffset;
+    const QgsPointXY pt = e->mapPoint();
+    const double XDistance = pt.x() - mStartPointMapCoords.x();
+    const double YDistance = pt.y() - mStartPointMapCoords.y();
+    double rotation = std::atan2( XDistance, YDistance ) * ( 180 / M_PI ) - mRotationOffset;
```
Comment (nyalldawson) [inline src/app/qgsmaptoolrotatefeature.cpp:167]: Why the X/y flip?

Comment (uclaros) [reply src/app/qgsmaptoolrotatefeature.cpp:167]: I'm not exactly sure, didn't investigate it. Switching from `e->pos()` screen coordinates to `e->mapPoint()` map coordinates reversed the rotation direction, so I flipped the atan.

Comment (nyalldawson) [reply src/app/qgsmaptoolrotatefeature.cpp:167]: Maybe a result a moving from y top to bottom painter coordinates to y bottom to top geographic coordinates?

Comment (uclaros) [reply src/app/qgsmaptoolrotatefeature.cpp:167]: That was what I suspected but wasn't sure if painter coords y axis was inverted.

**Comment by lbartoletti** [comment]
Thanks @uclaros ! Please, can you also add tests in [testqgsmaptoolrotatefeature.cpp](https://github.com/qgis/QGIS/blob/master/tests/src/app/testqgsmaptoolrotatefeature.cpp)

File: src/app/qgsmaptoolrotatefeature.cpp (hunk 1 above)
Comment (uclaros) [reply src/app/qgsmaptoolrotatefeature.cpp:167]: Is there a crs with an inverted y axis? Maybe I should test it.

**Comment by uclaros** [comment]
I sure will, this is still in draft!

File: src/app/qgsmaptoolrotatefeature.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/app/qgsmaptoolrotatefeature.cpp:167]: There are some south african ones, but QGIS doesn't support these well in general. See https://github.com/qgis/QGIS/issues/32391

File: tests/src/app/testqgsmaptoolrotatefeature.cpp [hunk 2]
```diff
@@ -137,8 +139,8 @@ void TestQgsMapToolRotateFeature::testRotateFeature()
   utils.mouseMove( 2, 1 );
   utils.mouseClick( 2, 1, Qt::LeftButton, Qt::KeyboardModifiers(), true );
 
-  QCOMPARE( mLayerBase->getFeature( 1 ).geometry().asWkt( 2 ), QStringLiteral( "Polygon ((0.72 -0.17, 0.28 1.17, 1.17 0.72, 0.72 -0.17))" ) );
-  QCOMPARE( mLayerBase->getFeature( 2 ).geometry().asWkt( 2 ), QStringLiteral( "Polygon ((1.1 0, 1.1 5, 2.1 5, 2.1 0))" ) );
+  QCOMPARE( mLayerBase->getFeature( 1 ).geometry().asWkt( 2 ), QStringLiteral( "Polygon ((-0.17 0.28, 0.28 1.17, 1.17 0.72, 0.72 -0.17, -0.17 0.28))" ) );
```
Comment (nyalldawson) [inline tests/src/app/testqgsmaptoolrotatefeature.cpp:142]: Why have these values changed?

Comment (uclaros) [reply tests/src/app/testqgsmaptoolrotatefeature.cpp:142]: The initial polygons wkt1 and wkt2 were invalid, they were open rings missing their last vertex. As a result, when those squares were rotated, a triangle was produced. In this commit I added the missing vertex to the polygons so the rotated geometries became squares too. I generally don't like modifying tests, but this one made no sense!

**Comment by uclaros** [comment]
@nyalldawson Since #42683 is getting a backport, should I also backport this one too?

**Comment by nyalldawson** [comment]
Yes, good idea

**Pull request merged**